Кравченко Юлии ПИ19-4

Задача теории игр

Вариант 12

In [1]:
import scipy
from scipy.optimize import linprog 
import numpy as np
import pandas as pd

Задание из Методички

In [70]:
c = [-1,-1,-1] # список коэффициентов функции цели
b_ub = [1,1,1] 
A_ub = [[1,2,3],
        [5,3,4],
        [1,4,0]]
d=linprog(c, A_ub, b_ub,method='simplex' ) 
#print(d)
for key,val in d.items():
         print(key,val) 
         if key=='x':
                  q=[sum(i) for i in A_ub*val]
                  print('A_ub*x',q) #использованные ресурсы
                  q1= scipy.array(b_ub)-scipy.array(q) 
                  print('b_ub-A_ub*x', q1) #остатки ресурсов

x [0.     0.25   0.0625]
A_ub*x [0.6875, 1.0, 1.0]
b_ub-A_ub*x [0.3125 0.     0.    ]
fun -0.3125
slack [0.3125 0.     0.    ]
con []
status 0
message Optimization terminated successfully.
nit 3
success True


C:\Users\kjd20\AppData\Local\Temp/ipykernel_1428/1864998524.py:13: DeprecationWarning: scipy.array is deprecated and will be removed in SciPy 2.0.0, use numpy.array instead
  q1= scipy.array(b_ub)-scipy.array(q)


Оптимальный план  x [0.     0.25   0.0625]

Оптимальное значение целевой функции fun -0.3125

In [72]:
#Оптимальные стратегии второго игрока и цена игры
w1_opt = 0
w2_opt = 0.25
w3_opt = 0.0625
F_max = 0.3125
SumW =d.x.sum()
y1 = w1_opt/SumW
y2 = w2_opt/SumW
y3 = w3_opt/SumW
Nu_A = 1/SumW # - цена игры
print("X* =",(y1,y2,y3),"-оптимальная стратегия  II игрока ")
print("Цена игры:", Nu_A)


X* = (0.0, 0.8, 0.2) -оптимальная стратегия  II игрока 
Цена игры: 3.2


In [4]:
#Двойственная задача - задача нахождения оптимальной стратегии для первого игрока
u1_opt = 0
u2_opt = 0.25
u3_opt = 0.0625
SumW =u1_opt+u2_opt+u3_opt
x1 = u1_opt/SumW
x2 = u2_opt/SumW
x3 = u3_opt/SumW

print("X* =",(x1,x2,x3),"-оптимальная стратегия  II игрока ")

X* = (0.0, 0.8, 0.2) -оптимальная стратегия  II игрока 


# Задания из ПЗ

Задача №1

In [5]:
df1 = pd.read_csv("lab8.csv", delimiter = ";", index_col=['n',"Телефоны","Наушники", "Ноутбуки"])
df1 = pd.DataFrame(df1)
df1

Empty DataFrame
Columns: []
Index: [(Телефоны, 300, 280, 230), (Ноутбуки, 200, 180, 130), (ПК, 240, 380, 400)]

Для подстановки :
lab8.csv

Телефоны Ноутбуки ПК

Телефоны Наушники Ноутбуки

300 280 230 200 180 130 240 380 400

In [5]:
def Z1(file):
    df1 = pd.read_csv(file, delimiter = ";",index_col='n')
    #print(df1) lab8.csv
    df1 = pd.DataFrame(df1)
    print(df1)
    optim1 = df1.min(axis=1)
    optim1w = pd.DataFrame(optim1)
    strat_strw = optim1w[optim1w[0]==optim1w[0].max()]
    print("1) Оптимальная чистая стратегия для игрока А и Цена игры для игрока А при выборе чистой оптимальной стратегии:",strat_strw)

    optim1B = df1.max()
    optim1B = pd.DataFrame(optim1B)
    stratB_str1 = optim1B[optim1B[0]==optim1B[0].min()]
    print("2) Оптимальная чистая стратегия для игрока Б и Цена игры для игрока Б при выборе чистой оптимальной стратегии:",stratB_str1)

    np1 = np.array(df1)

    c = [-1,-1,-1] # список коэффициентов функции цели
    b_ub = [1,1,1] 

    d=linprog(c, np1, b_ub,method='simplex' ) 
    zd1 = 1/d.x.sum()
    print("3) Цена игры для игрока А при выборе смешанной оптимальной стратегии:",   zd1)
    t_sme = round((d.x[0]/d.x.sum())*100)
    t_sme2 = round((d.x[1]/d.x.sum())*100)
    t_sme3 = round((d.x[2]/d.x.sum())*100)
    T_sme = [{'Телефоны':t_sme,'Ноутбуки':t_sme2, 'ПК':t_sme3}]
    T_sme1 = pd.DataFrame(T_sme)
    print("Таблица смешанных стратегий для игрока А:", "\n" , T_sme1)
def Z12(df1):
    optim1 = df1.min(axis=1)
    #print(optim1)
    optim1 = pd.DataFrame(optim1) 
    strat_str = optim1[optim1[0]==optim1[0].max()]
    print("1) Оптимальная чистая стратегия для игрока А и Цена игры для игрока А при выборе чистой оптимальной стратегии:",strat_str)
    optim1B = df1.max()
    print(optim1B)
    optim1B = pd.DataFrame(optim1B)
    stratB_str = optim1B[optim1B[0]==optim1B[0].min()]
    print("2) Оптимальная чистая стратегия для игрока Б и Цена игры для игрока Б при выборе чистой оптимальной стратегии:",stratB_str)
    np1 = np.array(df1)

    c = [-1,-1,-1] # список коэффициентов функции цели
    b_ub = [1,1,1] 
    d=linprog(c, np1, b_ub,method='simplex' )
    zd1 = 1/d.x.sum()
    print("3) Цена игры для игрока А при выборе смешанной оптимальной стратегии:", zd1)
    t_sme = round((d.x[0]/d.x.sum())*100)
    t_sme2 = round((d.x[1]/d.x.sum())*100)
    t_sme3 = round((d.x[2]/d.x.sum())*100)
    T_sme = [{'Телефоны':t_sme,'Ноутбуки':t_sme2, 'ПК':t_sme3}]
    T_sme1 = pd.DataFrame(T_sme)
    print("Таблица смешанных стратегий для игрока А:", "\n" ,T_sme1)

nums = []

inpt = input('Введите 1 если хотите взять данные из файл, 2 если хотите ввести к клавиатуры')



inpt = inpt.strip()
if inpt == '1':
    inpt1 = input('файл:')
    inpt1 = inpt1.strip()
    inpt1 = Z1(inpt1)
    
    

elif inpt == '2':
    S1 = input("Введите название 1 стратегии компании А :")
    S2 = input("Введите название 2 стратегии компании А :")
    S3 = input("Введите название 3 стратегии компании А :")
    S21 = input("Введите название 1 стратегии компании B :")
    S22 = input("Введите название 2 стратегии компании B :")
    S23 = input("Введите название 3 стратегии компании B :")
    a1,a2,a3,b1,b2,b3,c1,c2,c3 = input("Введите матрицу весовых коэффициентов поэлементно:").split()
    data1 = [{S1:a1,S2:a2, S3:a3},{S1:b1,S2:b2, S3:b3},{S1:c1,S2:c2, S3:c3}]
    df1 = pd.DataFrame(data1, index=[S21, S22,S23])
    Z12(df1)
else:
    print('Неправильное действие.\n')
            




1) Оптимальная чистая стратегия для игрока А и Цена игры для игрока А при выборе чистой оптимальной стратегии:               0
Ноутбуки  240.0
Телефоны    300
Ноутбуки    380
ПК          400
dtype: object
2) Оптимальная чистая стратегия для игрока Б и Цена игры для игрока Б при выборе чистой оптимальной стратегии:             0
Телефоны  300
3) Цена игры для игрока А при выборе смешанной оптимальной стратегии: 281.7391304347826
Таблица смешанных стратегий для игрока А: 
    Телефоны  Ноутбуки  ПК
0        74         0  26


Задание 2

Для подстановки 

Файлы:

lab8.csv    lab8.2_1.csv

Телефоны Ноутбуки ПК

Телефоны Наушники Ноутбуки

300 280 230 200 180 130 240 380 400

180 110 240 120 160 210 300 140 250

In [13]:
def Z2(file, file1):
    df2 = pd.read_csv(file, delimiter = ";",index_col='n')
    #print(df1) lab8.csv lab8.2_1.csv
    df2 = pd.DataFrame(df2)
    df22 = pd.read_csv(file1, delimiter = ";",index_col='n')
    df22 = pd.DataFrame(df22)
    print(df2)
    print(df22)
    optim1Bim = df2.max()
    
    optim1Bim = pd.DataFrame(optim1Bim)
    BiM_str = optim1Bim[optim1Bim[0]==optim1Bim[0].min()]
    print("1) Оптимальная чистая стратегия для игрока А и Цена игры для игрока А при выборе чистой оптимальной стратегии:", BiM_str)

    optim1Bim1 = df22.max(axis=1)
    
    optim1Bim1 = pd.DataFrame(optim1Bim1)
    BiM1_str = optim1Bim1[optim1Bim1[0]==optim1Bim1[0].min()]
    print("2) Оптимальная чистая стратегия для игрока Б и Цена игры для игрока А при выборе чистой оптимальной стратегии:", BiM1_str)

    ww = np.array(BiM1_str)
    ww1= np.array(BiM_str)
    print("3) Общая (суммарная) цена игры:",ww[0]+ww1[0])

    np2 = np.array(df2)
    
    c = [-1,-1,-1] # список коэффициентов функции цели
    b_ub = [1,1,1] 

    d2=linprog(c, np2, b_ub,method='simplex' ) 
    zd2 = 1/d2.x.sum()
    print("4) Цена игры для игрока А при выборе смешанной оптимальной стратегии:", zd2)

    #t_sme21 = round((d2.x[0]/d2.x.sum())*100)
    #t_sme22 = round((d2.x[1]/d2.x.sum())*100)
   ## t_sme23 = round((d2.x[2]/d2.x.sum())*100)
   # T_sme12 = [{'Телефоны':i,'Ноутбуки':i, 'ПК':i}]
    #T_sme21 = pd.DataFrame(T_sme12)
    #print("Таблица смешанных стратегий A:", "\n" , T_sme21)


    np2 = np.array(df2)
    np22 = np.array(df22)
    s = np2+np22
    d22=linprog(c, s, b_ub,method='simplex') 
    #print(d22)

    d3 = linprog(c, np2, b_ub,method='simplex') 
    #print(d3)
    t1_sme = (d3.x[0]/d3.x.sum())
    t1_sme2 = (d3.x[1]/d3.x.sum())
    t1_sme3 = (d3.x[2]/d3.x.sum())
    Q1 = [t1_sme,t1_sme2,t1_sme3]  
    #print("Q1",Q1)  
    i1 = d22.x[0]/d22.x.sum()
    i2 = 1-d22.x[1]/d22.x.sum()
    i3 = 1-d22.x[2]/d22.x.sum()
    W = [i1,i2,i3]
    Smesh = ((W *np22.T).T * Q1).sum()

    T_sme12 = [{'Телефоны':i1*100,'Ноутбуки':i2*100, 'ПК':i3*100}]
    T_sme21 = pd.DataFrame(T_sme12)
    print("5)Таблица смешанных стратегий A:", "\n" , T_sme21)
    print("6) Цена игры для игрока Б при выборе смешанной оптимальной стратегии:",Smesh)

    t_sme213 = round((d3.x[0]/d3.x.sum())*100)
    t_sme223 = round((d3.x[1]/d3.x.sum())*100)
    t_sme233 = round((d3.x[2]/d3.x.sum())*100)
    T_sme123 = [{'Телефоны':t_sme213,'Наушники':t_sme223, 'Ноутбуки':t_sme233}]
    T_sme213 = pd.DataFrame(T_sme123)
    print("7)Таблица смешанных стратегий B:", "\n" , T_sme213)

    #np22 = np.array(df22)
    #c = [-1,-1,-1] # список коэффициентов функции цели
    #b_ub = [1,1,1] 
    #d22=linprog(c, np22, b_ub,method='simplex' ) 
    #zd2 = 1/d22.x.sum()
    #print("5) Цена игры для игрока Б при выборе смешанной оптимальной стратегии:", zd2)

    sumOb = zd2+Smesh
    print("Общая цена игры в случае использования оптимальных стратегий:", sumOb)

def Z22(df2, df22):

    optim1Bim = df2.max()
    
    optim1Bim = pd.DataFrame(optim1Bim)
    BiM_str = optim1Bim[optim1Bim[0]==optim1Bim[0].min()]
    print("1) Оптимальная чистая стратегия для игрока А и Цена игры для игрока А при выборе чистой оптимальной стратегии:", BiM_str)

    optim1Bim1 = df22.max(axis=1)
    
    optim1Bim1 = pd.DataFrame(optim1Bim1)
    BiM1_str = optim1Bim1[optim1Bim1[0]==optim1Bim1[0].min()]
    print("2) Оптимальная чистая стратегия для игрока Б и Цена игры для игрока А при выборе чистой оптимальной стратегии:", BiM1_str)

    ww = int(np.array(BiM_str))
    ww1= int(np.array(BiM1_str))
    print("3) Общая (суммарная) цена игры:",ww+ww1)


    np2 = np.array(df2)
    
    c = [-1,-1,-1] # список коэффициентов функции цели
    b_ub = [1,1,1] 

    d2=linprog(c, np2, b_ub,method='simplex' ) 
    zd2 = 1/d2.x.sum()
    print("4) Цена игры для игрока А при выборе смешанной оптимальной стратегии:", zd2)

    #t_sme21 = round((d2.x[0]/d2.x.sum())*100)
    #t_sme22 = round((d2.x[1]/d2.x.sum())*100)
    #t_sme23 = round((d2.x[2]/d2.x.sum())*100)
    #T_sme12 = [{'Телефоны':t_sme21,'Ноутбуки':t_sme22, 'ПК':t_sme23}]
    #T_sme21 = pd.DataFrame(T_sme12)
    #print("Таблица смешанных стратегий:", "\n" , T_sme21)

    #np22 = np.array(df22)
    #c = [-1,-1,-1] # список коэффициентов функции цели
    #b_ub = [1,1,1] 
    ##d22=linprog(c, np22, b_ub, method='simplex' ) 
    #zd22 = 1/d22.x.sum()
    #print("5) Цена игры для игрока Б при выборе смешанной оптимальной стратегии:", zd22)

    #sumOb = zd2+zd22
    #print("Общая цена игры в случае использования оптимальных стратегий:", sumOb)


    np2 = np.array(df2)
    np22 = np.array(df22)
    #print(np2)
    sss = np.zeros((3,3))

    for i in range(3):
        for j in range(3):
            sss[i,j] = int(np22[i,j])
    sss1 = np.zeros((3,3))
    #print(sss1)

    for i in range(3):
        for j in range(3):
            sss1[i,j] = int(np2[i,j])        
            
    #print(sss)        
    s = sss1+sss
    d22=linprog(c, s, b_ub,method='simplex') 
    #print(d22)

    d3 = linprog(c, sss1, b_ub,method='simplex') 
    #print(d3)
    t1_sme = (d3.x[0]/d3.x.sum())
    t1_sme2 = (d3.x[1]/d3.x.sum())
    t1_sme3 = (d3.x[2]/d3.x.sum())
    Q1 = [t1_sme,t1_sme2,t1_sme3]  
    #print("Q1",Q1)  
    i1 = d22.x[0]/d22.x.sum()
    i2 = 1-d22.x[1]/d22.x.sum()
    i3 = 1-d22.x[2]/d22.x.sum()
    W = [i1,i2,i3]
    #print(W)
    Smesh = ((W *sss.T).T * Q1).sum()

    T_sme12 = [{'Телефоны':i1*100,'Ноутбуки':i2*100, 'ПК':i3*100}]
    T_sme21 = pd.DataFrame(T_sme12)
    print("5)Таблица смешанных стратегий A:", "\n" , T_sme21)
    print("6) Цена игры для игрока Б при выборе смешанной оптимальной стратегии:",Smesh)

    t_sme213 = round((d3.x[0]/d3.x.sum())*100)
    t_sme223 = round((d3.x[1]/d3.x.sum())*100)
    t_sme233 = round((d3.x[2]/d3.x.sum())*100)
    T_sme123 = [{'Телефоны':t_sme213,'Наушники':t_sme223, 'Ноутбуки':t_sme233}]
    T_sme213 = pd.DataFrame(T_sme123)
    print("7)Таблица смешанных стратегий B:", "\n" , T_sme213)

    #np22 = np.array(df22)
    #c = [-1,-1,-1] # список коэффициентов функции цели
    #b_ub = [1,1,1] 
    #d22=linprog(c, np22, b_ub,method='simplex' ) 
    #zd2 = 1/d22.x.sum()
    #print("5) Цена игры для игрока Б при выборе смешанной оптимальной стратегии:", zd2)

    sumOb = zd2+Smesh
    print("Общая цена игры в случае использования оптимальных стратегий:", sumOb)




nums = []

inpt = input('Введите 1 если хотите взять данные из файл, 2 если хотите ввести к клавиатуры')



inpt = inpt.strip()
if inpt == '1':
    inpt1 = input('файл 1 :')
    inpt2 = input('файл 2 :')
    inpt1 = inpt1.strip()
    inpt2 = inpt2.strip()
    inpt1 = Z2(inpt1, inpt2)
    
    
    
elif inpt == '2':
    S1 = input("Введите название 1 стратегии компании А :")
    S2 = input("Введите название 2 стратегии компании А :")
    S3 = input("Введите название 3 стратегии компании А :")
    S21 = input("Введите название 1 стратегии компании B :")
    S22 = input("Введите название 2 стратегии компании B :")
    S23 = input("Введите название 3 стратегии компании B :")
    a1,a2,a3,b1,b2,b3,c1,c2,c3 = input("Введите матрицу весовых коэффициентов А поэлементно:").split()
    a_1,a_2,a_3,b_1,b_2,b_3,c_1,c_2,c_3 = input("Введите матрицу весовых коэффициентов В поэлементно:").split()
    data2 = [{S1:a1,S2:a2, S3:a3},{S1:b1,S2:b2, S3:b3},{S1:c1,S2:c2, S3:c3}]
    df2 = pd.DataFrame(data2, index=[S1, S2,S3])
    data22 = [{S21:a_1,S22:a_2, S23:a_3},{S21:b_1,S22:b_2, S23:b_3},{S21:c_1,S22:c_2, S23:c_3}]
    df22 = pd.DataFrame(data22, index=[S21, S22,S23])
    Z22(df2, df22)
    
else:
    print('Неправильное действие.\n')
            


import csv

          Телефоны  Наушники  Ноутбуки
n                                     
Телефоны       300       280       230
Ноутбуки       200       180       130
ПК             240       380       400
          Телефоны  Наушники  Ноутбуки
n                                     
Телефоны       180       110       240
Наушники       120       160       210
Ноутбуки       300       140       250
1) Оптимальная чистая стратегия для игрока А и Цена игры для игрока А при выборе чистой оптимальной стратегии:             0
Телефоны  300
2) Оптимальная чистая стратегия для игрока Б и Цена игры для игрока А при выборе чистой оптимальной стратегии:             0
n            
Наушники  210
3) Общая (суммарная) цена игры: [510]
4) Цена игры для игрока А при выборе смешанной оптимальной стратегии: 281.7391304347826
5)Таблица смешанных стратегий A: 
    Телефоны  Ноутбуки     ПК
0       0.0       0.0  100.0
6) Цена игры для игрока Б при выборе смешанной оптимальной стратегии: 286.95652173913044
7)Таблица 

Задача 3

Для подстановки:

Файл:  lab8.3.csv 

Продукты Косметика Одежда Бытовая химия

Рецессия Стагнация Оживление Подъем 

Вероятности

50 30 35 45 40 40 55 40 35 40 45 55 10 20 25 35
0.15 0.25 0.5 0.1

In [8]:
def Z3(file):
    df3 = pd.read_csv(file, delimiter = ";",index_col='n')
    #print(df1) lab8.3.csv 
    df3 = pd.DataFrame(df3)
    print(df3)
    ss = np.array(df3)
   
    ##Критерий Лапласа
    ## Если вероятности состояний природы правдоподобны, для их оценки используют принцип 
    # недостаточного основания Лапласа, согласно которого все состояния природы полагаются равновероятными, т.е.:
    # q1 = q2 = ... = qn = 1/n.
    # qi = 1/4
    m = 4
    n = 4
    c_min = np.zeros((m,n))
    for i in range(m):
        for j in range(n):
            c_min[i, j] = (ss[i, j]*0.25)
            
    
    qwe = c_min.sum(axis=1)
    LAPL_df3 = pd.DataFrame(qwe, index=['Продукты', 'Косметика','Одежда', "Бытовая химия"] )
  
    LAPL_str = LAPL_df3[LAPL_df3[0]==LAPL_df3[0].max()]
    print("1) Оптимальная стратегия игрока А по критерию Лапласа и Цена игры с оптимальной стратегии по критерию Лапласа:", LAPL_str)

    ## Критерий Байеса
    for i in range(m):
        for j in range(n):
            c_min[i, j] = (ss[i, j]*ss[4,j])
        
  
    qwe = c_min.sum(axis=1)
    BAER_df3 = pd.DataFrame(qwe, index=['Продукты', 'Косметика','Одежда', "Бытовая химия"] )
    #print(BAER_df3)
    BAER_df3 = BAER_df3[BAER_df3[0]==BAER_df3[0].max()]
    print("2) Оптимальная стратегия игрока А по критерию Баеса и Цена игры с оптимальной стратегии по критерию Баеса:", BAER_df3)

    ##Критерий Гермейера
    c_min = np.zeros((m,n))
    for i in range(m):
        for j in range(n):
            c_min[i, j] = (ss[i, j]*ss[4,j])
        
    print(c_min)
    qwe = c_min.min(axis=1)
    Germe_df3 = pd.DataFrame(qwe, index=['Продукты', 'Косметика','Одежда', "Бытовая химия"] )
    print(Germe_df3)
    strG = Germe_df3[Germe_df3[0]==Germe_df3[0].max()]
    print("3) Оптимальная чистая стратегия игрока А по критерию Гермейера и цена игры:", strG)


def Z33(df3):
    ss = np.array(df3)
   
    ##Критерий Лапласа
    m = 4
    n = 4
    c_min = np.zeros((m,n))
    for i in range(m):
        for j in range(n):
            c_min[i,j] = int(ss[i,j])
            c_min[i, j] = c_min[i, j]*0.25
            
  
    qwe = c_min.sum(axis=1)
    LAPL_df3 = pd.DataFrame(qwe, index=['Продукты', 'Косметика','Одежда', "Бытовая химия"] )
    #print(LAPL_df3)
    LAPL_str = LAPL_df3[LAPL_df3[0]==LAPL_df3[0].max()]
    print("1) Оптимальная стратегия игрока А по критерию Лапласа и Цена игры с оптимальной стратегии по критерию Лапласа:", LAPL_str)

    ## Критерий Байеса
    m = 4
    n = 4
    m1 = 5
    n1=4
    sss = np.zeros((m1,n1))
    for i in range(m1):
        for j in range(n1):
            sss[i,j] = float(ss[i,j])
          
    c_min = np.zeros((m,n))
    for i in range(m):
        for j in range(n):
            c_min[i, j] = (sss[i, j]*sss[4,j])
        
  
    qwe = c_min.sum(axis=1)
    BAER_df3 = pd.DataFrame(qwe, index=['Продукты', 'Косметика','Одежда', "Бытовая химия"] )
    #print(BAER_df3)
    BAER_df3 = BAER_df3[BAER_df3[0]==BAER_df3[0].max()]
    print("2) Оптимальная стратегия игрока А по критерию Баеса и Цена игры с оптимальной стратегии по критерию Баеса:", BAER_df3)

    ##Критерий Гермейера
   
    m = 4
    n = 4
    m1 = 5
    n1=4
    sss = np.zeros((m1,n1))
    for i in range(m1):
        for j in range(n1):
            sss[i,j] = float(ss[i,j])
    #print(sss) 

    s  = np.all(np.matrix(sss) > 0)
    if s == True:
        print("Матрица Больше 0")

    c_min = np.zeros((m,n))
    for i in range(m):
        for j in range(n):
            c_min[i, j] = (sss[i, j]*sss[4,j])    

        
    #print(c_min)
    qwe = c_min.min(axis=1)
    Germe_df3 = pd.DataFrame(qwe, index=['Продукты', 'Косметика','Одежда', "Бытовая химия"] )
    #print(Germe_df3)
    strG = Germe_df3[Germe_df3[0]==Germe_df3[0].max()]
    print("3) Оптимальная чистая стратегия игрока А по критерию Гермейера и цена игры:", strG)
    


nums = []

inpt = input('Введите 1 если хотите взять данные из файл, 2 если хотите ввести к клавиатуры')



inpt = inpt.strip()
if inpt == '1':
    inpt1 = input('файл 1 :')
    #inpt2 = input('файл 2 :')
    inpt1 = inpt1.strip()
    #inpt2 = inpt2.strip()
    inpt1 = Z3(inpt1)
    
    
     #   print('Сумма введенных чисел:', rez)
elif inpt == '2':
    S1 = input("Введите название 1 стратегии компании А :")
    S2 = input("Введите название 2 стратегии компании А :")
    S3 = input("Введите название 3 стратегии компании А :")
    S4 = input("Введите название 4 стратегии компании А :")
    S21 = input("Введите название 1 состояний природы B :")
    S22 = input("Введите название 2 состояний природы B :")
    S23 = input("Введите название 3 состояний природы B :")
    S24 = input("Введите название 4 состояний природы B :")
    SS = input("Введите название Вероятности :")
    a1,a2,a3,a4,b1,b2,b3,b4,c1,c2,c3,c4,d1,d2,d3,d4 = input("Введите матрицу весовых коэффициентов поэлементно:").split()
    e1,e2,e3,e4 = input("Введите вектор вероятности:").split()
    data3 = [{S21:a1,S22:a2, S23:a3, S24:a4},{S21:b1,S22:b2, S23:b3, S24:b4},{S21:c1,S22:c2, S23:c3, S24:c4},{S21:d1,S22:d2, S23:d3, S24:d4},{S21:e1,S22:e2, S23:e3, S24:e4}]
    df3 = pd.DataFrame(data3, index=[S1,S2,S3,S4,SS])
    print(df3)
    Z33(df3)
    #print('Произведение введенных чисел:', rez)
else:
    print('Неправильное действие.\n')
            


import csv

               Регрессия  Стагнация  Оживление  Подъем
n                                                     
Продукты           50.00      30.00       35.0    45.0
Косметика          40.00      40.00       55.0    40.0
Одежда             35.00      40.00       45.0    55.0
Бытовая химия      10.00      20.00       25.0    35.0
Вероятность         0.15       0.25        0.5     0.1
1) Оптимальная стратегия игрока А по критерию Лапласа и Цена игры с оптимальной стратегии по критерию Лапласа:                0
Косметика  43.75
Одежда     43.75
2) Оптимальная стратегия игрока А по критерию Баеса и Цена игры с оптимальной стратегии по критерию Баеса:               0
Косметика  47.5
[[ 7.5   7.5  17.5   4.5 ]
 [ 6.   10.   27.5   4.  ]
 [ 5.25 10.   22.5   5.5 ]
 [ 1.5   5.   12.5   3.5 ]]
                  0
Продукты       4.50
Косметика      4.00
Одежда         5.25
Бытовая химия  1.50
3) Оптимальная чистая стратегия игрока А по критерию Гермейера и цена игры:            0
Одежда  5.25


Задача 4

Для подстановки

lab8.4.csv

Продукты Косметика Одежда Бытовая химия

Рецессия Стагнация Оживление Подъем 


50 30 35 45 40 40 40 40 70 40 35 50 10 20 25 35

In [73]:
def Z4(file):
    df4 = pd.read_csv(file, delimiter = ";",index_col='n')
    #print(df1)  lab8.4.csv
    df4 = pd.DataFrame(df4)
    ##Критерий Пессимизма
    pess = df4.min(axis=1)
    pess = pd.DataFrame(pess)
    pess_str = pess[pess[0]==pess[0].min()]
    print("1) Оптимальная стратегия игрока А по критерию пессимизма и цена игры:", pess_str[:])

    ## Критерий Оптимизма
    opt = df4.max(axis=1)
    opt = pd.DataFrame(opt)
    opt_str = opt[opt[0]==opt[0].max()]
    print("2) Оптимальная стратегия игрока А по критерию оптимизма и цена игры:", opt_str[:])

    ##Критерий Вальда
    vald = df4.min(axis=1)
    vald = pd.DataFrame(vald)
    vald_str = vald[vald[0]==vald[0].max()]
    print("3) Оптимальная стратегия игрока А по критерию Вальда и цена игры:", vald_str[:])

    ##Критерий Сэвиджа
    ss = np.array(df4)

    m = 4
    n = 4
    
    for i in range(m):
        for j in range(n):
            c_min21 = (ss[:,0].max()-ss[:,0])
            c_min22 = (ss[:,1].max()-ss[:,1])
            c_min23 = (ss[:,2].max()-ss[:,2])
            c_min24 = (ss[:,3].max()-ss[:,3])
    sevid = pd.DataFrame((c_min21,c_min22,c_min23,c_min24))
    sevid_transposed = sevid.T    

    qwe = sevid_transposed.max(axis=1)
    SEVID_df4 = pd.DataFrame(qwe)
    SEVID_df4a = SEVID_df4.rename(index={0: "Продукты", 1: "Косметика", 2: "Одежда",3:"Бытовая химия"})
    sev_str = SEVID_df4a[SEVID_df4a[0]==SEVID_df4a[0].min()]
    print("4) Оптимальная стратегия игрока А по критерию Сэвиджа и Величина минимальной недополученной прибыли по критерию Сэвиджа:", sev_str)
    


def Z44(df4):
    ##Критерий Пессимизма
    pess = df4.min(axis=1)
    pess = pd.DataFrame(pess)
    pess_str = pess[pess[0]==pess[0].min()]
    print("1) Оптимальная стратегия игрока А по критерию пессимизма и цена игры:", pess_str)

    ## Критерий Оптимизма
    opt = df4.max(axis=1)
    opt = pd.DataFrame(opt)
    opt_str = opt[opt[0]==opt[0].max()]
    print("2) Оптимальная стратегия игрока А по критерию оптимизма и цена игры:", opt_str)

    ##Критерий Вальда
    vald = df4.min(axis=1)
    vald = pd.DataFrame(vald)
    vald_str = vald[vald[0]==vald[0].max()]
    print("3) Оптимальная стратегия игрока А по критерию Вальда и цена игры:", vald_str)

    ##Критерий Сэвиджа
    ss = np.array(df4)
    #print(ss)

    m = 4
    n = 4
    ss1 = np.zeros((m,n))
    for i in range(m):
        for j in range(n):
            ss1[i,j] = int(ss[i,j])
            #ss1 = map(int, ss)
            #print(ss1)
            c_min21 = (ss1[:,0].max()-ss1[:,0])
            c_min22 = (ss1[:,1].max()-ss1[:,1])
            c_min23 = (ss1[:,2].max()-ss1[:,2])
            c_min24 = (ss1[:,3].max()-ss1[:,3])
    sevid = pd.DataFrame((c_min21,c_min22,c_min23,c_min24))
    sevid_transposed = sevid.T    
    #print(sevid_transposed)
    qwe = sevid_transposed.max(axis=1)
    SEVID_df4 = pd.DataFrame(qwe)
    SEVID_df4a = SEVID_df4.rename(index={0: "Продукты", 1: "Косметика", 2: "Одежда",3:"Бытовая химия"})
    #print(SEVID_df4a)
    sev_str = SEVID_df4a[SEVID_df4a[0]==SEVID_df4a[0].min()]
    print("4) Оптимальная стратегия игрока А по критерию Сэвиджа и Величина минимальной недополученной прибыли по критерию Сэвиджа:", sev_str)

nums = []

inpt = input('Введите 1 если хотите взять данные из файл, 2 если хотите ввести к клавиатуры')



inpt = inpt.strip()
if inpt == '1':
    inpt1 = input('файл 1 :')
    inpt1 = inpt1.strip()
    inpt1 = Z4(inpt1)
    
    
     #   print('Сумма введенных чисел:', rez)
elif inpt == '2':
    S1 = input("Введите название 1 стратегии компании А :")
    S2 = input("Введите название 2 стратегии компании А :")
    S3 = input("Введите название 3 стратегии компании А :")
    S4 = input("Введите название 4 стратегии компании А :")
    S21 = input("Введите название 1 состояния :")
    S22 = input("Введите название 2 состояния :")
    S23 = input("Введите название 3 состояния :")
    S24 = input("Введите название 4 состояния :")
    a1,a2,a3,a4,b1,b2,b3,b4,c1,c2,c3,c4,d1,d2,d3,d4 = input("Введите матрицу весовых коэффициентов поэлементно:").split()
    data4 = [{S21:a1,S22:a2, S23:a3,S24:a4},{S21:b1,S22:b2, S23:b3,S24:b4},{S21:c1,S22:c2, S23:c3,S24:c4},{S21:d1,S22:d2, S23:d3,S24:d4}]
    df4 = pd.DataFrame(data4, index=[S1, S2,S3,S4])
    
    Z44(df4)
    #print('Произведение введенных чисел:', rez)
else:
    print('Неправильное действие.\n')
            


import csv

1) Оптимальная стратегия игрока А по критерию пессимизма и цена игры:                 0
n                
Бытовая химия  10
2) Оптимальная стратегия игрока А по критерию оптимизма и цена игры:          0
n         
Одежда  70
3) Оптимальная стратегия игрока А по критерию Вальда и цена игры:             0
n            
Косметика  40
4) Оптимальная стратегия игрока А по критерию Сэвиджа и Величина минимальной недополученной прибыли по критерию Сэвиджа:         0
Одежда  5
